### ISSUE TITLE
Parse TypeScript syntax

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#690]
Parser and variable analyzer:

- [x] interfaces
- [x] classes
- [x] type annotations on variable declarations
- [x] `<type>variable` type assertions
- [x] abstract methods
- [x] `as` type assertions
- [x] `!` postfix operator
- [x] type expression syntax
- [x] generic parameters
- [x] generic arguments
- [x] decorators (#964)
- [x] enums
- [x] `import type`
- [x] `export interface`, `export type`, etc.
- [x] namespaces
- [x] `declare`
- [x] `module`
- [x] type aliases
- [x] `this` parameters
- [x] [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates)
- [x] predeclared global types (e.g. `Generator`)
- [x] `type T<U>=number;` (`>=`) #848
- [x] `export abstract class`
- [x] [`satisfies`](https://devblogs.microsoft.com/typescript/announcing-typescript-4-9-beta/)
- [x] `infer` (e.g. https://github.com/evanw/esbuild/issues/2712)
- [x] #919
- [x] #921
- [x] generic `in`/`out`
- [x] `constructor(public foo: number)`, `constructor(public readonly foo: number)`
- [x] `function f<const T>() {}`
- [x] TODOs (e.g. #974, #981)
- [x] `public declare readonly myField: number;`
- [x] `export declare function ...`
- [x] `<Type<GenericArg> />` (JSX + generic arguments)
- [x] .d.ts syntax
- [ ] #1107
- [x] `let foo!: Bar;`

Integrations:

- [x] CLI: `.ts`, `.tsx`, `.d.ts`
- [x] CLI: `--language=typescript` option
- [x] LSP
- [x] VS Code extension
- [x] Vim plugins
- [ ] Emacs plugins
  - [x] Flymake
  - [ ] Flycheck #1148
  - [ ] Eglot #1146
  - [ ] lsp-mode #1147
- [x] Sublime Text plugin
- [ ] Kate plugin #1149
- [x] web demo
- [x] Experimental:
  - [x] CLI: `--language=experimental-typescript` option
  - [x] CLI: `.ts`, `.tsx`, `.d.ts` (`--language=experimental-default` option)
  - [x] LSP
  - [ ] Emacs plugins
  - [x] VS Code extension
  - [x] Vim plugins

**Comment by strager** [comment]
TypeScript keywords: https://github.com/microsoft/TypeScript/blob/99ffa394a838a5528d6118fb3c23f81fd8e99ab2/src/compiler/scanner.ts#L81

**Comment by strager** [comment]
# Interfaces

* `interface`s are typechecking-only and cannot be used in expressions.
* `function`s, `const`/`let`/`var`/`catch` variables, and parameters are runtime-only and cannot be used in types.
* `class`es are usable both in expressions and in types.
* `import`ed variables are usable according to the above rules. Without looking at the imported module, there is no way to know if an imported variable can be used in an expression or in a type.

Conflicts:

* If an interface and a function/variable have the same name, there is no ambiguity.
* If an interface has the same name as an imported `class` or `interface`, there is a compile error.
* If an interface has the same name as an imported `function`/`const`/`let`/`var`, there is no compile error.
* If an interface and a class have the same name, then the declarations seem to be *merged* (union of all members). This can lead to violations in the type system. For example:
```typescript
interface I { field: number }
class I { }
let o = new I(); // not a type error
console.log(o.field.toString()); // run-time error!
```

**Comment by strager** [comment]
Some interesting syntax quirks related to JSX: https://stackoverflow.com/a/54614279/39992

**Comment by strager** [comment]
Here's an interesting file we could use for performance testing (almost 7 MiB): https://raw.githubusercontent.com/NativeScript/NativeScript/main/packages/types-android/src/lib/android/android-platform-31.d.ts

**Comment by strager** [comment]
TypeScript has too many syntax-level features. @_@

**Comment by vegerot** [comment]
> TypeScript has too many syntax-level features. @_@

you're tellin' me!

**Comment by strager** [comment]
Shipped in version 3.0.0.
